miner/worker.go

```diff
@@ -360,619 +238,200 @@ func (w *worker) createTxCmp() func(tx1 *types.Transaction, tx2 *types.Transacti
-	w.current.receipts = append(w.current.receipts, receipt)
-
-	return receipt.Logs, nil
-}
-
-func (w *worker) commitTransactions(txs *types.TransactionsByPriceAndNonce, txFeeRecipient common.Address, interrupt *int32) bool {
-	// Short circuit if current is nil
-	if w.current == nil {
-		return true
-	}
-
-	if w.current.gasPool == nil {
-		w.current.gasPool = new(core.GasPool).AddGas(w.current.gasLimit)
-	}
+// mainLoop is a standalone goroutine to create tasks and submit to the engine.
+func (w *worker) mainLoop() {
+	defer w.chainHeadSub.Unsubscribe()
+	defer w.txsSub.Unsubscribe()
+	// Context and cancel function for the currently executing block construction
+	// Cancel needs to be called in each exit path to make the linter happy
+	// because go struggles with analyzing lexical scoping.
+	var taskCtx context.Context
+	var cancel context.CancelFunc
 
-	var coalescedLogs []*types.Log
+	txsCh := make(chan core.NewTxsEvent, txChanSize)
 
-loop:
-	for {
-		// In the following three cases, we will interrupt the execution of the transaction.
-		// (1) new head block event arrival, the interrupt signal is 1
-		// (2) worker start or restart, the interrupt signal is 1
-		// (3) worker recreate the mining block with any newly arrived transactions, the interrupt signal is 2.
-		// For the first two cases, the semi-finished work will be discarded.
-		// For the third case, the semi-finished work will be submitted to the consensus engine.
-		if interrupt != nil && atomic.LoadInt32(interrupt) != commitInterruptNone {
-			// Notify resubmit loop to increase resubmitting interval due to too frequent commits.
-			if atomic.LoadInt32(interrupt) == commitInterruptResubmit {
-				ratio := float64(w.current.gasLimit-w.current.gasPool.Gas()) / float64(w.current.gasLimit)
-				if ratio < 0.1 {
-					ratio = 0.1
-				}
-				w.resubmitAdjustCh <- &intervalAdjust{
-					ratio: ratio,
-					inc:   true,
-				}
-			}
-			return atomic.LoadInt32(interrupt) == commitInterruptNewHead
-		}
-		// If we don't have enough gas for any further transactions then we're done
-		if w.current.gasPool.Gas() < params.TxGas {
-			log.Trace("Not enough gas for further transactions", "have", w.current.gasPool, "want", params.TxGas)
-			break
-		}
-		// Retrieve the next transaction and abort if all done
-		tx := txs.Peek()
-		if tx == nil {
-			break
-		}
-		// Short-circuit if the transaction requires more gas than we have in the pool.
-		// If we didn't short-circuit here, we would get core.ErrGasLimitReached below.
-		// Short-circuiting here saves us the trouble of checking the GPM and so on when the tx can't be included
-		// anyway due to the block not having enough gas left.
-		if w.current.gasPool.Gas() < tx.Gas() {
-			log.Trace("Skipping transaction which requires more gas than is left in the block", "hash", tx.Hash(), "gas", w.current.gasPool.Gas(), "txgas", tx.Gas())
-			txs.Pop()
-			continue
+	newWorkFunc := func() {
+		if cancel != nil {
```

Comment:
It's important to note that `cancel()` doesn't imply the goroutine is finished, just that it letting it know that it should finish. 

It would be nice if we wait for the routinte to finish, so we use Context as the kill signal, and a WaitGroup to wait for it to finish.

That way we know for a fact there are not two concurrent goroutines trying to build a block at the same time